OpenTelemetry JDBC instrumentation - fix Oracle and DB2 in native mode

michalvavrik opened the pull request:
fixes: #28915
fixes: #30176

Oracle driver is not registered with DriverManager in native mode. When Oracle driver is used directly, e.g. by [Agroal here](https://github.com/agroal/agroal/blob/master/agroal-pool/src/main/java/io/agroal/pool/ConnectionFactory.java#L123) it's not a problem, however [OpenTelemetry driver uses](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/OpenTelemetryDriver.java#L103) `DriverManager` and thus the Oracle driver is missing. I wonder if we should initialize it at runtime, however [there is a warning](https://github.com/quarkusio/quarkus/blob/main/extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java#L75) in `OracleMetadataOverrides` that deferred me from doing it. Instead, this PR register driver when `quarkus-oracle-jdbc` extension is used together with OpenTelemetry extension. Way I understand `java.sql.DriverManager#registerDriver(java.sql.Driver)` implementation, it's simply ignored if the driver is already there, however it would be nice to find better way in the future (or now if you have one!).

DB2 driver wasn't registered for reflection when used with OT driver as then Agroal registered OT driver and not DB2 driver in native mode.

michalvavrik commented:
Could you have a look @Sanne, please?

michalvavrik commented:
(Windows) failures are not related.

Sanne commented:
Interesting, unfortuntaly my options are limited as the Oracle driver is not open source.

In general I would suggest to not rely on the `DriverManager` - such static global registries are very ancient code patterns, I'm sure we can do better today. We're not relying on `DriverManager` in other code in Quarkus (as far as I know - clearly I didn't know about OpenTelemetry using it) and actively trying to not trigger it.

Would that be possible? It also happens to solve this issue, but I also believe it would be better overall.

michalvavrik commented:
OpenTelemetry driver is relying on DriverManager https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/OpenTelemetryDriver.java#L103 and do not offer other way to set drivers, so there are 2 ways we can work around this:

- byte code transformation of either one of drivers (Oracle/ OT)
- forcing runtime initialization of Oracle driver (didn't do that according to your comments)

Registering Oracle driver for now and opening issue in OT (yet to do) felt like the easiest way to work around the issue.

geoand commented:
I am also -1 for this unless there is absolutely no other way to go about it.

@radcortez @brunobat can't we contribute something to OTel around this?

Sanne commented:
This might indeed be our only option though.

Anyone knows why it's a problem specific for Oracle and why we shuouldn't register all drivers on `DriverManager` is OTel requires it?

michalvavrik commented:
> Anyone knows why it's a problem specific for Oracle and why we shuouldn't register all drivers on `DriverManager` if OTel requires it?

Adjusted reproducer for Postgresql and the issue is not there even though `-H:+PrintClassInitialization` shows `org.postgresql.Driver, BUILD_TIME`. I can't find direct call to `org.postgresql.Driver#register` except in static init block & PGBundleActivator (in pgjdbc project, can't see anything in quarkus/ot projects either so far). Don't have the answer, but can't see how there could be other explanation than that something calls `org.postgresql.Driver#register`.

michalvavrik commented:
Hey @brunobat & @radcortez ,

I hope I didn't block solution with this PR - e.g. hope you don't wait for my. Here is my suggestion: I'll open an issue in OpenTelemetry and also provide a PR that close it. The issue should allow us to register drivers without `java.sql.DriverManager#registerDriver(java.sql.Driver)`.

I propose it takes a while to get this fixed in OT and get it released, so it would be nice to have workaround till then. This PR does that, should you prefer different one, please, bring it on!

brunobat commented:
Hi @michalvavrik, 
Thanks for the PR. I was on PTO, therefore the delay.
I agree on having this provisional solution on, with some tweaks. Will provide review comments next.
If you want to provide a link for the OTel side PR, I'm happy to review and help with it.

brunobat commented:
> > Hi @michalvavrik, Thanks for the PR. I was on PTO, therefore the delay. I agree on having this provisional solution on, with some tweaks. The PR looks good to me but later, will the final solution we should move this instrumentation wiring elsewhere and create specific test projects per DB...
> 
> Yeah, but that will mean 3 new IT modules, right? (fine with me, just checking).
> 
> > We should assert also do PostgreSQL and MariaDB, at least.

We can go with TestContainers and have just one JDBC Driver related project with different test classes. One different test resource for each test class: https://quarkus.io/guides/getting-started-testing#quarkus-test-resource


> 
> I checked manually https://github.com/michalvavrik/quarkus-native-oracle-otel-reproducer and PG and MariaDB are fine, which confuses me heavily as they all (including OracleDriver) works pretty same way - register themselves in static initializer block. Not sure what's different and will try to identify exact reason, but avoiding DriverManager still seems like a good way around this (with OT PR).
> 
> > If you want to provide a link for the OTel side PR, I'm happy to review and help with it.
> 
> Thank you, I opened an issue [open-telemetry/opentelemetry-java-instrumentation#7028](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/7028) and will draft a PR over weekend and ask you for review/kindly provide help if I go in a wrong direction etc.

michalvavrik commented:
> We can go with TestContainers and have just one JDBC Driver related project with different test classes. One different test resource for each test class: https://quarkus.io/guides/getting-started-testing#quarkus-test-resource

Thanks, done.

brunobat commented on integration-tests/opentelemetry-jdbc-instrumentation/README.md:
````diff
@@ -0,0 +1,16 @@
+To also test as a native image, add `-Dnative`:
+
+```
+mvn verify -Dtest-containers -Dnative
````
Are you missing the details about the property you need to set to run a specific DB?

michalvavrik replied:
I added information how to set Oracle/PostgreSQL/MariaDB to a specific version. Thank you for the suggestion.

brunobat reviewed (commented):
This looks really good @michalvavrik. Thanks very much. 
Left a small comment.

michalvavrik commented:
I'll check test failure later today.

gsmet commented on integration-tests/pom.xml:
```diff
@@ -310,6 +310,7 @@
                 <module>micrometer-mp-metrics</module>
                 <module>micrometer-prometheus</module>
                 <module>opentelemetry</module>
+                <module>opentelemetry-jdbc-instrumentation</module>
```
AFAICS, the original issue was with native executables and I don't see the module added here: https://github.com/quarkusio/quarkus/blob/main/.github/native-tests.json ?

Is it an oversight or your intention?

michalvavrik replied:
I forgot, it should be fixed now (I'll check CI), thanks.

gsmet reviewed (changes requested):
Let's wait for @michalvavrik 's feedback before merging.

michalvavrik commented:
I think fact that `OracleOpenTelemetryJdbcInstrumentationTest.testOracleQueryTraced` failed (this time, not last time...) on JDK 18 and nowhere else shows that this test is going to be flaky as it can't be ensured Oracle container will start within one minute (hang limit enforced by CI). I don't think we can somehow "hurry" it up.

radcortez commented:
@michalvavrik were you able to determine why this works without changes with other drivers, and not with the Oracle one? Ideally, we should make sure that all drivers we support work with OTel in native mode.

michalvavrik commented:
> @michalvavrik were you able to determine why this works without changes with other drivers, and not with the Oracle one?

I didn't. Only difference I can see is that in static initializer block of the OracleDriver, there are multiple calls to other methods, f.e. to `registerMBeans` that is substituted by `quarkus-jdbc-oracle` in order to make this work. I run native build with GraalVM debug options I could find and there is nothing interesting.

> Ideally, we should make sure that all drivers we support work with OTel in native mode.

You can only expect it to work with JDBC drivers that Quarkus prepared to work in native mode (pg, oracle, mysql, mariadb, mssql, h2, derby db2). I didn't see runtime initializations in other JDBC extension, but only way to be sure is to have integration tests, while Oracle JDBC extensions won't run without them. Alternatively, there is a PR to OT instrumentation and if it ever gets accepted, we can simply register all drivers directly.

If you have an idea how to get more information / direction where to look, please let me know.

Tomorrow I'll move Oracle IT from testcontainer to docker container run by plugin as I don't think Quarkus CI will raise hang timeout just because of my test.

radcortez commented:
> You can only expect it to work with JDBC drivers that Quarkus prepared to work in native mode (pg, oracle, mysql, mariadb, mssql, h2, derby db2).

Yes, I was referring to the drivers we support. I know there are no issues with the pg one, but if we have issues with the oracle one, maybe other may also have it.

michalvavrik commented:
> 

I created reproducer for all our drivers here https://github.com/michalvavrik/quarkus-native-oracle-otel-reproducer, it seems only Oracle and DB2 don't work OOTB.

DB2 didn't work as when it's not registered by Agroal extension for reflection, class is removed (because Agroal registers OT driver instead), so I added registration for reflection and now it works.

As for "knowing" - I added 3 ITs as that's what Bruno wanted (MariaDB, Oralce, PostgreSQL), now that I was fixing DB2 I also prepared one for DB2 but not committed (can add the test quickly if you want), it's necessary to realize these db containers may start slowly and it is going to eat resources and time. F.e. DB2 container starts 2 minute on my workstation, so CI with 1 minute hanging limit will be reached again.

michalvavrik commented:
fyi @gsmet JDK 18 test is failing as root cause is `##[warning]You are running out of disk space. The runner will stop working when the machine runs out of disk space. Free space left: 77 MB`. The lack of space is before tests started, so I don't think I can fix that. Have you seen it before? Suggestions?

brunobat commented:
@michalvavrik About DB2... If it takes that much time to execute it should not run. However it is relevant code to have around and if you want to commit it as a deactivated test, it would be nice.

radcortez commented:
> > Yes, I was referring to the drivers we support. I know there are no issues with the pg one, but if we have issues with the oracle one, maybe others may also have it.
> 
> I created reproducer for all our drivers here https://github.com/michalvavrik/quarkus-native-oracle-otel-reproducer, it seems only Oracle and DB2 don't work OOTB.
> 
> DB2 didn't work as when it's not registered by Agroal extension for reflection, class is removed (because Agroal registers OT driver instead), so I added registration for reflection and now it works.
> 
> As for "knowing" - I added 3 ITs as that's what Bruno wanted (MariaDB, Oralce, PostgreSQL), now that I was fixing DB2 I also prepared one for DB2 but not committed (can add the test quickly if you want), it's necessary to realize these db containers may start slowly and it is going to eat resources and time. F.e. DB2 container starts 2 minute on my workstation, so CI with 1 minute hanging limit will be reached again.

Awesome work! Thank you!

brunobat commented:
@michalvavrik @gsmet Are you guys ok with this issue now? 
Can this be approved and merged?

michalvavrik commented:
@brunobat I kinda forget about it as I was very busy in last weeks. https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7089 got merged so it will be present in 1.20.3, until then we can merge this as workaround. Tomorrow I'll have a look how to work around CI limitations, it's bit out of my hands as it fails over resources (small memory), but I'll try to come with something...

brunobat commented:
Take your time @michalvavrik, no worries.

michalvavrik commented:
btw @brunobat 1.21.0 is here https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v1.21.0

brunobat commented:
@michalvavrik Please check: https://github.com/quarkusio/quarkus/pull/29863

gastaldi commented on integration-tests/opentelemetry-jdbc-instrumentation/README.md:
````diff
@@ -0,0 +1,36 @@
+mvn verify -Dtest-containers -Dstart-containers -Dnative
+```
+
+You can also run tests with a specific database image
+You can also run tests with a specific database image, just set:
````
```suggestion
You can also run tests with a specific database image, just set the following parameters:
```

gastaldi commented on integration-tests/opentelemetry-jdbc-instrumentation/README.md:
````diff
@@ -0,0 +1,37 @@
+mvn verify -Dtest-containers -Dstart-containers -Dnative
+```
+
+You can also run tests with a specific database image
````
Delete this line

michalvavrik replied:
I completely missed it @gastaldi , thank you.

gastaldi commented on integration-tests/opentelemetry-jdbc-instrumentation/README.md:
````diff
@@ -0,0 +1,37 @@
+mvn verify -Dtest-containers -Dstart-containers -Dpostgres.image=docker.io/postgres:latest
+```
+
+Unfortunately booting DB2 is slow, needs to set a generous timeout, therefore the DB2 test is disabled by default.
````
```suggestion
Unfortunately booting DB2 is slow and needs to set a generous timeout, therefore the DB2 test is disabled by default.
```

michalvavrik committed:
OpenTelemetry JDBC instrumentation - fix Oracle and DB2 in native mode

michalvavrik commented:
I've seen these in unrelated PRs before, they are flaky tests.

michalvavrik commented:
Hello @jerboaa ,
you [added registration for reflection few days back](https://github.com/quarkusio/quarkus/commit/5199b62303a53bbaeb47e94444be4c1a552f7a73#diff-4e56842dfe3e884c6ef5b7abe0a47bb995437353b87356f0ee20a26c9b167a69R34) in https://github.com/quarkusio/quarkus/pull/30012 of `com.sun.jndi.dns.DnsContextFactory` and `com.sun.jndi.rmi.registry.RegistryContextFactory` into `jdbc-oracle` extension, which breaks native build in this PR. What is your take on this failure, please? cc reviewer @zakkak 

I reverted your changes locally and native build works as expected.

```bash
========================================================================================================================
GraalVM Native Image: Generating 'quarkus-integration-test-opentelemetry-jdbc-instrumentation-999-SNAPSHOT-runner' (executable)...
========================================================================================================================
Warning: Feature class oracle.nativeimage.NativeImageFeature is annotated with the deprecated annotation @AutomaticFeature. Support for this annotation will be removed in a future version of GraalVM. Applications should register a feature using the option --features=oracle.nativeimage.NativeImageFeature
[1/7] Initializing...                                                                                   (18.8s @ 0.49GB)
 Version info: 'GraalVM 22.3.0.1-Final Java 17 Mandrel Distribution'
 Java version info: '17.0.5+8'
 C compiler: gcc (redhat, x86_64, 8.5.0)
 Garbage collector: Serial GC
 10 user-specific feature(s)
 - com.oracle.svm.thirdparty.gson.GsonFeature
 - io.quarkus.caffeine.runtime.graal.CacheConstructorsFeature
 - io.quarkus.hibernate.orm.runtime.graal.DisableLoggingFeature: Disables INFO logging during the analysis phase for the [org.hibernate.Version, org.hibernate.annotations.common.Version, org.hibernate.dialect.Dialect] categories
 - io.quarkus.jdbc.postgresql.runtime.graal.SQLXMLFeature
 - io.quarkus.runner.Feature: Auto-generated class by Quarkus from the existing extensions
 - io.quarkus.runtime.graal.DisableLoggingFeature: Disables INFO logging during the analysis phase for the [org.jboss.threads] categories
 - io.quarkus.runtime.graal.ResourcesFeature: Register each line in META-INF/quarkus-native-resources.txt as a resource on Substrate VM
 - oracle.nativeimage.NativeImageFeature
 - org.hibernate.graalvm.internal.GraalVMStaticFeature: Hibernate ORM's static reflection registrations for GraalVM
 - org.hibernate.graalvm.internal.QueryParsingSupport: Hibernate ORM's support for HQL Parser in GraalVM
The bundle named: com/sun/rowset/RowSetResourceBundle, has not been found. If the bundle is part of a module, verify the bundle name is a fully qualified class name. Otherwise verify the bundle path is accessible in the classpath.
12:49:23,763 WARN  [io.qua.ope.run.exp.otl.LateBoundBatchSpanProcessor] No BatchSpanProcessor delegate specified, no action taken.
[2/7] Performing analysis...  [*]                                                                       (76.9s @ 3.61GB)
  19,756 (90.12%) of 21,923 classes reachable
  29,612 (55.49%) of 53,365 fields reachable
  94,428 (58.35%) of 161,826 methods reachable
     781 classes,   590 fields, and 3,538 methods registered for reflection

Fatal error: com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing sun.rmi.transport.LiveRef.<init>(int) 
Parsing context:
   at sun.rmi.transport.LiveRef.<init>(LiveRef.java:74)
   at sun.rmi.server.UnicastServerRef.<init>(UnicastServerRef.java:166)
   at java.rmi.server.UnicastRemoteObject.exportObject(UnicastRemoteObject.java:360)
   at java.rmi.server.UnicastRemoteObject.<init>(UnicastRemoteObject.java:232)
   at java.rmi.server.UnicastRemoteObject.<init>(UnicastRemoteObject.java:214)

	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:153)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.createFlowsGraph(MethodTypeFlow.java:104)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureFlowsGraphCreated(MethodTypeFlow.java:83)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.getOrCreateMethodFlowsGraph(MethodTypeFlow.java:65)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.typestate.DefaultSpecialInvokeTypeFlow.onObservedUpdate(DefaultSpecialInvokeTypeFlow.java:61)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:562)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.PointsToAnalysis$1.run(PointsToAnalysis.java:488)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:193)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:177)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: org.graalvm.compiler.java.BytecodeParser$BytecodeParserError: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap. Instances created during image generation have cached seed values and don't behave as expected.  To see how this object got instantiated use --trace-object-instantiation=java.security.SecureRandom. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
	at parsing java.rmi.server.ObjID.<init>(ObjID.java:119)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.throwParserError(BytecodeParser.java:2518)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.throwParserError(SharedGraphBuilderPhase.java:110)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3393)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.handleBytecodeBlock(BytecodeParser.java:3345)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3190)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1138)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1030)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:97)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:84)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:446)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.AnalysisParsedGraph.parseBytecode(AnalysisParsedGraph.java:135)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisMethod.ensureGraphParsed(AnalysisMethod.java:685)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.phases.InlineBeforeAnalysisGraphDecoder.lookupEncodedGraph(InlineBeforeAnalysis.java:180)
	at jdk.internal.vm.compiler/org.graalvm.compiler.replacements.PEGraphDecoder.doInline(PEGraphDecoder.java:1162)
	at jdk.internal.vm.compiler/org.graalvm.compiler.replacements.PEGraphDecoder.tryInline(PEGraphDecoder.java:1145)
	at jdk.internal.vm.compiler/org.graalvm.compiler.replacements.PEGraphDecoder.trySimplifyInvoke(PEGraphDecoder.java:1003)
	at jdk.internal.vm.compiler/org.graalvm.compiler.replacements.PEGraphDecoder.handleInvoke(PEGraphDecoder.java:957)
	at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.GraphDecoder.processNextNode(GraphDecoder.java:817)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.phases.InlineBeforeAnalysisGraphDecoder.processNextNode(InlineBeforeAnalysis.java:240)
	at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.GraphDecoder.decode(GraphDecoder.java:548)
	at jdk.internal.vm.compiler/org.graalvm.compiler.replacements.PEGraphDecoder.decode(PEGraphDecoder.java:833)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.phases.InlineBeforeAnalysis.decodeGraph(InlineBeforeAnalysis.java:98)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:179)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:349)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.createFlowsGraph(MethodTypeFlow.java:93)
	... 13 more
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap. Instances created during image generation have cached seed values and don't behave as expected.  To see how this object got instantiated use --trace-object-instantiation=java.security.SecureRandom. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.DisallowedImageHeapObjectFeature.error(DisallowedImageHeapObjectFeature.java:174)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.image.DisallowedImageHeapObjects.check(DisallowedImageHeapObjects.java:65)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.DisallowedImageHeapObjectFeature.replacer(DisallowedImageHeapObjectFeature.java:150)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisUniverse.replaceObject(AnalysisUniverse.java:595)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.replaceObject(AnalysisConstantReflectionProvider.java:177)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.interceptValue(AnalysisConstantReflectionProvider.java:148)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readValue(AnalysisConstantReflectionProvider.java:100)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readFieldValue(AnalysisConstantReflectionProvider.java:79)
	at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.util.ConstantFoldUtil$1.readValue(ConstantFoldUtil.java:55)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.common.spi.JavaConstantFieldProvider.readConstantField(JavaConstantFieldProvider.java:78)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ameta.AnalysisConstantFieldProvider.readConstantField(AnalysisConstantFieldProvider.java:72)
	at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.util.ConstantFoldUtil.tryConstantFold(ConstantFoldUtil.java:51)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.ConstantFoldLoadFieldPlugin.tryConstantFold(ConstantFoldLoadFieldPlugin.java:53)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.ConstantFoldLoadFieldPlugin.handleLoadStaticField(ConstantFoldLoadFieldPlugin.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genGetStatic(BytecodeParser.java:4801)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genGetStatic(BytecodeParser.java:4772)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5282)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3385)
	... 38 more
``

jerboaa commented:
> Hello @jerboaa , you [added registration for reflection few days back](https://github.com/quarkusio/quarkus/commit/5199b62303a53bbaeb47e94444be4c1a552f7a73#diff-4e56842dfe3e884c6ef5b7abe0a47bb995437353b87356f0ee20a26c9b167a69R34) in #30012 of `com.sun.jndi.dns.DnsContextFactory` and `com.sun.jndi.rmi.registry.RegistryContextFactory` into `jdbc-oracle` extension, which breaks native build in this PR. What is your take on this failure, please?

It looks like with this additional reflection config, this application finds some RMI classes being reachable and thus, fails when it tries to build-time-initialize a `SecureRandom`. We could conditionalize on the used graal vm version, but that doesn't seem like it would be the correct fix. Do you know what is pulling in those RMI classes? 

> I reverted your changes locally and native build works as expected.

I'm afraid, this would likely be the wrong fix.

I'll try to have a look.

michalvavrik commented:
> ```shell
> ========================================================================================================================
> ```

Thanks for having a look! I didn't mean revert as fix, just to show that before your commit this particular scenario was okay (while what you fixed wasn't).

michalvavrik commented:
> Do you know what is pulling in those RMI classes?

No, sorry I didn't have time yet. If you need me to investigate, let me know. AFAICT this PR changes nothing relevant to jdbc-oracle, it just creates instance of Oracle driver at runtime.

jerboaa commented:
> > Do you know what is pulling in those RMI classes?
> 
> No, sorry I didn't have time yet. If you need me to investigate, let me know. AFAICT this PR changes nothing relevant to jdbc-oracle, it just creates instance of Oracle driver at runtime.

There must be something different, though as with other native integration tests we don't see it. Do other tests not create the (Oracle) driver at runtime?

michalvavrik commented:
> There must be something different, though as with other native integration tests we don't see it. Do other tests not create the (Oracle) driver at runtime?

I think Agroal does that, but here new instance is created via reflection and there are no imports, how could it have effect for native build? There is one difference to other tests though - integration test that is failing has 4 jdbc dependencies: `quarkus-jdbc-oracle`, `quarkus-jdbc-postgresql`, `quarkus-jdbc-mariadb`, `quarkus-jdbc-db2`, that must be unusual.

jerboaa commented:
> `quarkus-jdbc-oracle`, `quarkus-jdbc-postgresql`, `quarkus-jdbc-mariadb`, `quarkus-jdbc-db2`, that must be unusual.

Yes, indeed.

jerboaa commented:
Seems to originate from some `SPNEGO` (possibly from some of the drivers):

```
   at com.sun.jndi.rmi.registry.RegistryContext.getRegistry(RegistryContext.java:418)
   at com.sun.jndi.rmi.registry.RegistryContext.<init>(RegistryContext.java:103)
   at com.sun.jndi.url.rmi.rmiURLContext.getRootURLContext(rmiURLContext.java:324)
   at com.sun.jndi.toolkit.url.GenericURLContext.lookup(GenericURLContext.java:217)
   at com.sun.jndi.url.rmi.rmiURLContextFactory.getUsingURL(rmiURLContextFactory.java:71)
   at com.sun.jndi.url.rmi.rmiURLContextFactory.getObjectInstance(rmiURLContextFactory.java:56)
   at com.sun.jndi.rmi.registry.RegistryContextFactory.URLsToObject(RegistryContextFactory.java:115)
   at com.sun.jndi.rmi.registry.RegistryContextFactory.getObjectInstance(RegistryContextFactory.java:90)
   at javax.naming.spi.NamingManager.getURLObject(NamingManager.java:625)
   at javax.naming.spi.NamingManager.getURLContext(NamingManager.java:574)
   at sun.security.krb5.KrbServiceLocator.getKerberosService(KrbServiceLocator.java:85)
   at sun.security.krb5.Config.checkRealm(Config.java:1321)
   at sun.security.krb5.Config.getRealmFromDNS(Config.java:1294)
   at sun.security.krb5.Config.getDefaultRealm(Config.java:1187)
   at sun.security.krb5.Realm.getDefault(Realm.java:62)
   at sun.security.krb5.PrincipalName.<init>(PrincipalName.java:464)
   at sun.security.jgss.krb5.Krb5NameElement.getInstance(Krb5NameElement.java:121)
   at sun.security.jgss.krb5.Krb5MechFactory.getNameElement(Krb5MechFactory.java:107)
   at sun.security.jgss.GSSManagerImpl.getNameElement(GSSManagerImpl.java:193)
   at sun.security.jgss.GSSNameImpl.getElement(GSSNameImpl.java:472)
   at sun.security.jgss.GSSNameImpl.init(GSSNameImpl.java:202)
   at sun.security.jgss.GSSNameImpl.<init>(GSSNameImpl.java:171)
   at sun.security.jgss.GSSManagerImpl.createName(GSSManagerImpl.java:109)
   at sun.net.www.protocol.http.spnego.NegotiatorImpl.init(NegotiatorImpl.java:95)
   at sun.net.www.protocol.http.spnego.NegotiatorImpl.<init>(NegotiatorImpl.java:123)
```

jerboaa commented:
@michalvavrik https://github.com/michalvavrik/quarkus/pull/48 fixes the issue for me.

michalvavrik commented:
> @michalvavrik [michalvavrik#48](https://github.com/michalvavrik/quarkus/pull/48) fixes the issue for me.

Thanks, help really appreciated! I can see you added it to OT module, I'm really interested whether I could reproduce without OT, but I'm busy today. Let me have a look next week, or if you submit Quarkus PR, I will test it ex post.

jerboaa commented:
I'll let you run with it.

michalvavrik committed:
Fix native mode build when `jdbc-oracle` is combined with `jdbc-db2`

Fixes native build when `jdbc-oracle` is combined with `jdbc-db2`. This fix was kindly provided Severin Gehwolf and is added as second commit of this PR as 1. integration tests need the fix 2. no point of creating additional integration test module that combines jdbc-oracle and jdbc-db2 as we already have them here.

michalvavrik commented:
fyi @jerboaa - the issue is not related to OT, but comes from combination of `jdbc-oracle` and `jdbc-db2`, I don't want to create addition integration test modules as this PR already adds tests that verifies fix, so I included your fix into this PR as second commit with description. And thanks.

michalvavrik commented:
Quarkus QE experienced exactly same native build failures as described above, logs are here https://github.com/quarkus-qe/quarkus-test-suite/actions/runs/3814140863/jobs/6488338890; I verified it's fixed by this PR too.

I won't open an issue as there is a chance that @brunobat will review this PR and we can have it fixed in no time; in case that's a false hope, I'll open an issue later this week.

brunobat commented:
I checked out the PR and executed the db tests in my M1, in native mode and they pass... 
This must be a CI issue.

michalvavrik commented:
> I checked out the PR and executed the db tests in my M1, in native mode and they pass... This must be a CI issue.

I think so, thank you for checking. Please review the PR (or merge it) when times is right for you. I'm sure you've got a lot to catch up after holidays period. Thank you very much.

geoand commented:
What's the status of this one?

michalvavrik commented:
> What's the status of this one?

I guess we are waiting for additional reviewers.

brunobat commented:
@gsmet needs to review it

gsmet commented on extensions/jdbc/jdbc-db2/deployment/src/main/java/io/quarkus/jdbc/db2/deployment/JDBCDB2Processor.java:
```diff
@@ -51,6 +54,15 @@ void configureAgroalConnection(BuildProducer<AdditionalBeanBuildItem> additional
         }
     }
 
+    @BuildStep
+    void registerDriverForReflection(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        //Not strictly necessary when using Agroal, as it also registers
+        //any JDBC driver being configured explicitly through its configuration.
+        //We register it for the sake of people not using Agroal,
+        //for example when the driver is used with OpenTelemetry JDBC instrumentation.
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, DB2_DRIVER_CLASS));
+    }
+
```
I don't understand that part. Because I'm pretty sure that would be a problem for all the drivers around? Or is Db2 specific in this regard?

michalvavrik replied:
Adding this registering for reflection fixes the issue with OT as then Agroal does not register driver itself (it registers OT driver for reflection) and IT I added failed.

gsmet replied:
Okay :).

I think we should probably take a step back and implement OpenTelemetry for JDBC a bit differently by using a similar pattern to what we did in https://github.com/quarkusio/quarkus/pull/26144 for OpenTracing. I have not checked if it's doable but I think we should at least have a look.

@brunobat does it sound like a good idea to you?

In the meantime, let's merge this to fix the issue.

brunobat replied:
At the consistency level it would be preferable and it's also more user friendly.
What do you think @michalvavrik ?

brunobat replied:
created this: https://github.com/quarkusio/quarkus/issues/30532

michalvavrik replied:
Sorry, I'm very busy these days. I'll look into #30532 .

gsmet reviewed (commented):
I added one small question for something I don't understand fully. Could you check?

michalvavrik commented:
Rebased the PR to resolve merge conflicts. No changes made, still waiting for the review.
